### ISSUE TITLE
cpu 100%

### ISSUE THREAD

**Issue by chalmagean** [magnars/smart-forward.el#4]
Hey, I just tried `M-x smart-forward` in a ruby buffer and it seems to cause `GNU Emacs 24.2.1 (x86_64-apple-darwin, NS apple-appkit-1038.36) of 2012-08-27 on bob.porkrind.org` to take the cpu to 100%.

the code looks like this:

```
class SomeClass < ActiveRecord::Base
  [cursor_here]before_filter :foo
  ...
end
```

**Comment by magnars** [comment]
Thanks for the report, I'll have a look.

**Comment by jdahm** [comment]
Hi. First of all, I'm a big fan of emacsrocks, multiple-cursors, expand-region, and now this. Keep up the great work!

I'm also experiencing this issue. The only way way I've run into it so far is by editing a (fairly large) source file (I'm editing C right now), starting at the top of the file execute (smart-forward) and (smart-backward) a few times. I haven't had success debugging this since it just seems to hang and I'm not familiar with the expand-region internals. My thinking is that this could just be the algorithm expand-region is using taking too long. For complicated source files it could be doing too much processing.

For reference I'm using GNU Emacs 24.4.1.

Cheers, Johann

**Comment by magnars** [comment]
Yeah, expand-region uses specific expansion functions for specific modes, and can get extremely slow if one or more of those are slow. I'm guessing that some C-specific expansions aren't performing well with large files. If you want to send me such a file, I'll happily take a quick look to see if I can find the culprit.

**Comment by jdahm** [comment]
I took a look at the c-expansions and although there is indeed no formal tests or expansion support for comments, it's doing a surprisingly good job with them. Still, I think this could be the problem, since this code uses a fair number of C style /\* ... */ comments.

I've started a branch 'c-expansions' in my fork of expand-region.el that adds a test for this. However I don't know the first thing about cucumber testing and I have way too much work to get done lately, so this will probably have to wait a while. The branch is nowhere near ready for a pull request, but if you'd like to take a look and incorporate bits that would be excellent.

Also, I've added another branch 'c-failure' to my fork that adds another file (param.c - grafted from another project) that exemplifies the 100% cpu behavior. I stripped a lot from the file so fyi it definitely won't compile, if you were interested in that. I just thought this would be the easiest way to get you a quick example. Here's a direct link:
https://github.com/jdahm/expand-region.el/blob/c-failure/param.c
